src/klein/_plating.py

```diff
@@ -25,17 +35,74 @@ def _should_return_json(request):
     return bool(request.args.get(b"json"))
 
 
-def json_serialize(item):
+@inlineCallbacks
+def resolveDeferredObjects(root):
     """
-    A function similar to L{dumps}.
+    Wait on possibly nested L{Deferred}s that represent a JSON
+    serializable object.
+
+    @param root: A JSON-serializable object that may contain
+        L{Deferred}s, or a Deferred that will resolve to a
+        JSON-serializable object
+
+    @return: A L{Deferred} that fires with a L{Deferred}-free version
+        of C{root}, or that fails with the first exception
+        encountered.
     """
-    def helper(unknown):
-        if isinstance(unknown, PlatedElement):
-            return unknown._asJSON()
+
+    result = [None]
+    setResult = partial(setitem, result, 0)
+    stack = [(root, setResult)]
+
+    while stack:
+        mightBeDeferred, setter = stack.pop()
+        obj = yield mightBeDeferred
+        if isinstance(obj, ATOM_TYPES):
+            setter(obj)
+        elif isinstance(obj, list):
+            parent = [None] * len(obj)
+            setter(parent)
+            stack.extend(
+                reversed([
+                    (child, partial(setitem, parent, i))
+                    for i, child in enumerate(obj)
+                ])
+            )
+        elif isinstance(obj, tuple):
+            parent = [None] * len(obj)
+            setter(tuple(parent))
+
+            def setTupleItem(i, value, parent=parent, setter=setter):
+                parent[i] = value
+                setter(tuple(parent))
+
+            stack.extend(
+                reversed([
+                    (child, partial(setTupleItem, i))
+                    for i, child in enumerate(obj)
+                ])
+            )
+        elif isinstance(obj, dict):
+            parent = {}
+            setter(parent)
+            for key, value in reversed(list(obj.items())):
+                pair = [None, None]
+                setKey = partial(setitem, pair, 0)
+
+                def setValue(value, pair=pair, parent=parent):
+                    pair[1] = value
+                    parent.update([pair])
+
+                stack.append((value, setValue))
+                stack.append((key, setKey))
+        elif isinstance(obj, PlatedElement):
+            stack.append((obj._asJSON(), setter))
         else:
             raise TypeError("{input} not JSON serializable"
```

Comment:
`resolveDeferredObjects` should be extendable to arbitrary objects.

This could call a `default` function that accepts the object and returns something that's serializable.  The result and setter function would then get pushed back onto the stack, [like what happens with PlatedElement._asJSON](https://github.com/twisted/klein/pull/238/files#diff-43c63257bab6e48d2c5f6dc06f495020R98).

Reply:
Yeah, I was going to say, the builtin json.JSONEncoder has a default argument. This probably isn't the PR for it, but if klein ever gets more opinionated, would be nice to see datetime having some sane isoformat/isoparse handling.

Reply:
Nit: I like it when exceptions include the "perp" as an argument. That's really nice in debuggers.